File: zhangrui-api/src/main/java/com/zhangrui/study/algorithms/array/MajorityElement.java
package com.zhangrui.study.algorithms.array;

import java.util.Arrays;
import java.util.HashMap;
import java.util.Map;

/**
 * @Author: ZhangRui
 * @Date: Created at 2019-07-25-13:02
 * @Description: Given an array of size n, find the majority element.
 * The majority element is the element that appears more than ⌊ n/2 ⌋ times.
 * @Modified: By
 */
public class MajorityElement {

	public static int majorityElement(int[] nums) {
		if (null == nums || nums.length < 1) {
			throw new NullPointerException("The array is NULL!");
		}
		int length = nums.length;
		Integer zero = 0;
		Map<Integer, Integer> elements = new HashMap<>();
		for (int num : nums) {
			Integer count = elements.get(num);
			if (null == count || zero.equals(count)) {
				elements.put(num, 1);
				if (1 == length) {
					return num;
				}
			}else {
				elements.put(num, count + 1);
				if (elements.get(num) > (length / 2)) {
					return num;
				}
			}
		}
		throw new IllegalArgumentException("The array is illegal!");
	}

	public static int majorityElementAdvanced(int[] nums) {
		if (null == nums || nums.length < 1) {
			throw new NullPointerException("The array is NULL!");
		}
		Arrays.sort(nums);
		return nums[nums.length / 2];
	}

	public static void main(String[] args) {
		System.out.println(majorityElement(new int[]{1,2,1,2,2,0,0,0,0,0,0}));
		System.out.println(majorityElementAdvanced(new int[]{1,2,1,2,2,0,0,0,0,0,0}));
	}
}

File: PyQuest/algorithm/array/majority_element.py
class Solution:
    def majorityElement(self, nums: List[int]) -> int:
        # 我们想寻找的那个众数
        target = 0
        # 计数器（类比带电粒子例子中的带电性）
        count = 0
        for i in range(len(nums)):
            if count == 0:
                # 当计数器为 0 时，假设 nums[i] 就是众数
                target = nums[i]
                # 众数出现了一次
                count = 1
            elif nums[i] == target:
                # 如果遇到的是目标众数，计数器累加
                count += 1
            else:
                # 如果遇到的不是目标众数，计数器递减
                count -= 1
        # 回想带电粒子的例子
        # 此时的 count 必然大于 0，此时的 target 必然就是目标众数
        return target